prov/efa/src/rdm/rxr_ep.h

```diff
@@ -279,6 +286,78 @@ static inline int rxr_need_sas_ordering(struct rxr_ep *ep)
 	return ep->msg_order & FI_ORDER_SAS;
 }
 
+static inline uint64_t rxr_get_host_id(char *host_id_file)
+{
+	FILE *fp = NULL;
+	char *host_id_str = NULL, *parsed_host_id_str = NULL;
+	const size_t HOST_ID_LENGTH = 16;
+	int ret = 0;
+	long length = 0;
+	size_t read_length = 0;
+	uint64_t tmp_host_id = 0, host_id = 0;
+
+	if (!host_id_file) {
+		EFA_WARN(FI_LOG_EP_CTRL, "Host id file is not specified\n");
+		goto out;
+	}
+
+	fp = fopen(host_id_file, "r");
+	if (!fp) {
+		EFA_WARN(FI_LOG_EP_CTRL, "Cannot open host id file: %s\n", host_id_file);
+		goto out;
+	}
+
+	for (char c = fgetc(fp); c != '\n' && c != EOF; c = fgetc(fp)) {
+		++length;
+	}
+
+	/* Check negative value first because length is signed */
+	if (length < 0) {
+		EFA_WARN(FI_LOG_EP_CTRL, "Cannot get host id length: %d\n", errno);
+		goto out;
+	}
+
+	if (length < HOST_ID_LENGTH) {
+		EFA_WARN(FI_LOG_EP_CTRL, "Host id is too short. Expect: %ld bytes, got: %ld bytes\n", HOST_ID_LENGTH, length);
+		goto out;
+	}
+
+	fseek(fp, length - HOST_ID_LENGTH, SEEK_SET);
+
+	host_id_str = calloc(HOST_ID_LENGTH, sizeof(*host_id_str));
+	parsed_host_id_str = calloc(HOST_ID_LENGTH, sizeof(*parsed_host_id_str));
+	if (!host_id_str || !parsed_host_id_str) {
+		EFA_WARN(FI_LOG_EP_CTRL, "Failed to allocate memory\n");
+		goto out;
+	}
+
+	read_length = fread(host_id_str, sizeof(*host_id_str), HOST_ID_LENGTH, fp);
+	if (read_length != HOST_ID_LENGTH) {
```

Comment:
This seems like a lot of logic to ensure there are a constant number of hex characters in the file.

Why not just use a 16-byte array, read 16 bytes, and fail if you read less or they are invalid characters?  If you really want to confirm the file is not too long, read one more character and assert it's either non-hex or EOF.

Reply:
Yes that's my purpose. We want to read the last 16 characters - for that reason we need to find out the end of the first line and walk back 16 chars, then read from there.

> if they are invalid characters
Do you mean we test if the characters are 0-9/a-f? I can do that as well, but I'm not sure if that makes the logic simpler.

Reply:
@lrbison I attempted a simpler version, could you see if it's better or could be made better?

Reply:
@lrbison Thanks for your suggestion. I adopted your solution with minor tweaks. We still need the alloc.